pkg/dbhash/dbhash.go

```diff
@@ -0,0 +1,120 @@
+package dbhash
+
+import (
+	"context"
+	"crypto/sha1"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"io"
+)
+
+// DatabaseStateHash calculates the hash of some state of the database acorrding to options passed.
+func DatabaseStateHash(ctx context.Context, tx *sql.Tx, opts ...Option) (string, error) {
+	config := DefaultConfig()
+	for _, o := range opts {
+		if err := o(config); err != nil {
+			return "", fmt.Errorf("applying provided option: %s", err)
+		}
+	}
+
+	h := sha1.New()
+	if err := databaseStateWriter(ctx, tx, h, config); err != nil {
+		return "", fmt.Errorf("database hash writer failed: %s", err)
+	}
+
+	return hex.EncodeToString(h.Sum(nil)), nil
+}
+
+func databaseStateWriter(ctx context.Context, tx *sql.Tx, writer io.Writer, c *Config) error {
+	// get all tables from db and associated schema
+	rows, err := tx.QueryContext(ctx, c.FetchSchemasQuery)
```

Comment:
The query to fetch all tables is configurable. But there's a caveat: It kind of expects the query to have the columns (`tbl_name` and `sql`) that come from `sqlite_schema` table to the whole thing to work properly. There's not a validation for that, it kind of expects the client to do the right thing.